File: #0020.valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {

        map<char, char> determine{{')', '('}, {']', '['}, {'}', '{'}};
        stack<char> sc;
        for (char c : s) {
            if (!determine.count(c)) {
                sc.push(c);
            } else {
                if (sc.empty() || determine[c] != sc.top()) {
                    return false;
                }
                sc.pop();
            }
        }
        return sc.empty();
    }
};

File: #0020.valid-parentheses.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {

        let determine: HashMap<char, char> = [
            ('(', ')'), ('[', ']'), ('{', '}'),
        ].iter().cloned().collect();
        let mut sc: Vec<char> = vec![];
        for c in s.chars() {
            if determine.contains_key(&c) {
                sc.push(c)
            } else if sc.pop().and_then(|k| determine.get(&k)).unwrap_or(&'#') != &c { 
                return false;
            }
        }
        sc.is_empty()
    }
}
